# Logestic Regression Model - SFS

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sqlalchemy import create_engine
from config import db_password
import matplotlib.pyplot as plt
import seaborn as sns
import math

In [2]:
db_string = f"postgresql://postgres:{db_password}@127.0.0.1:5433/nba_champs"
engine = create_engine(db_string)

In [3]:
stats_df = pd.read_sql("select * from \"stats\"", db_string);
stats_df.head()

index  Position   Season   Tm   G   W   L  Win_pct   MOV   SOS  ...  \
0      0         0  2020-21  WAS  72  34  38    0.472 -1.83 -0.01  ...   
1      1         0  2020-21  UTA  72  52  20    0.722  9.25 -0.29  ...   
2      2         0  2020-21  TOR  72  27  45    0.375 -0.47 -0.07  ...   
3      3         0  2020-21  SAS  72  33  39    0.458 -1.74  0.15  ...   
4      4         0  2020-21  SAC  72  31  41    0.431 -3.68  0.23  ...   

   eFG_pct  TS_pct  TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TS_pct  \
0    0.531   0.569     12.3     21.3   0.221        0.539       0.576   
1    0.563   0.597     12.7     24.5   0.195        0.507       0.537   
2    0.529   0.567     11.9     20.8   0.196        0.543       0.584   
3    0.517   0.554     10.2     20.0   0.192        0.541       0.570   
4    0.549   0.578     12.0     21.3   0.185        0.557       0.591   

   Opp_TOV_pct  Opp_ORB_pct  Opp_FT_FGA  
0         12.5         22.4       0.217  
1         10.3         20.7       0.159  
2         14.4         23.7       0.234  
3         11.8         22.7       0.174  
4         12.1         25.0       0.199  

[5 rows x 24 columns]

In [4]:
test_df= pd.read_sql("select * from \"cy_stats\"", db_string)
test_df.head()

index  Position   Season   Tm  G  W  L  Win_pct    MOV   SOS  ...  eFG_pct  \
0      0         0  2021-22  MIA  6  5  1    0.833  17.00  1.07  ...    0.525   
1      1         0  2021-22  WAS  7  5  2    0.714   2.57  1.11  ...    0.505   
2      2         0  2021-22  UTA  6  5  1    0.833  12.83 -3.08  ...    0.526   
3      3         0  2021-22  TOR  8  5  3    0.625   4.63  1.25  ...    0.487   
4      4         0  2021-22  SAS  7  2  5    0.286  -0.43 -1.50  ...    0.526   

   TS_pct  TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TS_pct  Opp_TOV_pct  \
0   0.565     13.6     27.5   0.197        0.464       0.501         13.4   
1   0.549     11.5     20.4   0.213        0.492       0.535         11.0   
2   0.574     14.2     24.6   0.236        0.450       0.485         10.8   
3   0.518     12.1     29.8   0.158        0.521       0.554         16.8   
4   0.542     12.2     22.2   0.113        0.549       0.568         15.2   

   Opp_ORB_pct  Opp_FT_FGA  
0         15.9       0.181  
1         20.5       0.197  
2         23.1       0.164  
3         20.6       0.178  
4         21.4       0.130  

[5 rows x 24 columns]

In [5]:
# Pull out Season, TM and Team_Name into DF
name_df = stats_df[['Season', 'Tm',]]
name_df.head()

Season   Tm
0  2020-21  WAS
1  2020-21  UTA
2  2020-21  TOR
3  2020-21  SAS
4  2020-21  SAC

In [6]:
#list(stats_df.columns)

In [7]:
all_features_df = stats_df
all_features_df.head()

index  Position   Season   Tm   G   W   L  Win_pct   MOV   SOS  ...  \
0      0         0  2020-21  WAS  72  34  38    0.472 -1.83 -0.01  ...   
1      1         0  2020-21  UTA  72  52  20    0.722  9.25 -0.29  ...   
2      2         0  2020-21  TOR  72  27  45    0.375 -0.47 -0.07  ...   
3      3         0  2020-21  SAS  72  33  39    0.458 -1.74  0.15  ...   
4      4         0  2020-21  SAC  72  31  41    0.431 -3.68  0.23  ...   

   eFG_pct  TS_pct  TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TS_pct  \
0    0.531   0.569     12.3     21.3   0.221        0.539       0.576   
1    0.563   0.597     12.7     24.5   0.195        0.507       0.537   
2    0.529   0.567     11.9     20.8   0.196        0.543       0.584   
3    0.517   0.554     10.2     20.0   0.192        0.541       0.570   
4    0.549   0.578     12.0     21.3   0.185        0.557       0.591   

   Opp_TOV_pct  Opp_ORB_pct  Opp_FT_FGA  
0         12.5         22.4       0.217  
1         10.3         20.7       0.159  
2         14.4         23.7       0.234  
3         11.8         22.7       0.174  
4         12.1         25.0       0.199  

[5 rows x 24 columns]

In [8]:
stats_df.head()

index  Position   Season   Tm   G   W   L  Win_pct   MOV   SOS  ...  \
0      0         0  2020-21  WAS  72  34  38    0.472 -1.83 -0.01  ...   
1      1         0  2020-21  UTA  72  52  20    0.722  9.25 -0.29  ...   
2      2         0  2020-21  TOR  72  27  45    0.375 -0.47 -0.07  ...   
3      3         0  2020-21  SAS  72  33  39    0.458 -1.74  0.15  ...   
4      4         0  2020-21  SAC  72  31  41    0.431 -3.68  0.23  ...   

   eFG_pct  TS_pct  TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TS_pct  \
0    0.531   0.569     12.3     21.3   0.221        0.539       0.576   
1    0.563   0.597     12.7     24.5   0.195        0.507       0.537   
2    0.529   0.567     11.9     20.8   0.196        0.543       0.584   
3    0.517   0.554     10.2     20.0   0.192        0.541       0.570   
4    0.549   0.578     12.0     21.3   0.185        0.557       0.591   

   Opp_TOV_pct  Opp_ORB_pct  Opp_FT_FGA  
0         12.5         22.4       0.217  
1         10.3         20.7       0.159  
2         14.4         23.7       0.234  
3         11.8         22.7       0.174  
4         12.1         25.0       0.199  

[5 rows x 24 columns]

In [9]:
# Drop un needed columns
all_features_df.drop(['index',
 'Season',
 'Tm',
 'G',
 'W',
 'L',
 #'Win_pct',
 #'MOV',
 #'SOS',
 #'SRS',
 #'Pace',
 #'ORtg',
 #'DRtg',
 #'eFG_pct',
 #'TS_pct',
 #'TOV_pct',
 #'ORB_pct',
 #'FT_FGA',
 #'Opp_eFG_pct',
 #'Opp_TS_pct',
 #'Opp_TOV_pct',
 #'Opp_ORB_pct',
 #'Opp_FT_FGA'
              ], axis=1, inplace=True)

In [10]:
all_features_df.head()

Position  Win_pct   MOV   SOS   SRS   Pace   ORtg   DRtg  eFG_pct  TS_pct  \
0         0    0.472 -1.83 -0.01 -1.85  104.1  111.2  113.0    0.531   0.569   
1         0    0.722  9.25 -0.29  8.97   98.5  117.6  108.3    0.563   0.597   
2         0    0.375 -0.47 -0.07 -0.54   99.2  112.0  112.5    0.529   0.567   
3         0    0.458 -1.74  0.15 -1.58   98.9  111.0  112.8    0.517   0.554   
4         0    0.431 -3.68  0.23 -3.45  100.0  113.6  117.2    0.549   0.578   

   TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TS_pct  Opp_TOV_pct  \
0     12.3     21.3   0.221        0.539       0.576         12.5   
1     12.7     24.5   0.195        0.507       0.537         10.3   
2     11.9     20.8   0.196        0.543       0.584         14.4   
3     10.2     20.0   0.192        0.541       0.570         11.8   
4     12.0     21.3   0.185        0.557       0.591         12.1   

   Opp_ORB_pct  Opp_FT_FGA  
0         22.4       0.217  
1         20.7       0.159  
2         23.7       0.234  
3         22.7       0.174  
4         25.0       0.199

In [11]:
# Remove Position target from features data
y = all_features_df.Position
X = all_features_df.drop(columns=["Position"])

# Split training/test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)


In [12]:
feature_names = X.columns

In [13]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Scale the data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)

In [14]:
# Create a logistic regression model
classifier_AF = LogisticRegression(max_iter=1000)
classifier_AF

LogisticRegression(max_iter=1000)

In [15]:
# Fit (train) our model by using the training data
classifier_AF.fit(X_train_scaled, y_train)

LogisticRegression(max_iter=1000)

In [16]:
# Validate the model by using the test data
print(f"Logistic Regression Training Data Score: {classifier_AF.score(X_train_scaled, y_train)}")
print(f"Logistic Regression Testing Data Score: {classifier_AF.score(X_test_scaled, y_test)}")

Logistic Regression Training Data Score: 0.9642346208869814
Logistic Regression Testing Data Score: 0.9786324786324786


All features score = 0.97863

Find accuracy using SFM(SelectFromModel)

In [17]:
# https://towardsdatascience.com/5-feature-selection-method-from-scikit-learn-you-should-know-ed4d116e4172
from sklearn.feature_selection import SelectFromModel
smf=SelectFromModel(classifier_AF)
smf.fit(X_train, y_train)
X.columns[smf.get_support()]

Index(['Win_pct', 'MOV', 'SOS', 'SRS', 'ORtg', 'DRtg', 'FT_FGA',
       'Opp_TOV_pct'],
      dtype='object')

In [18]:
# Drop un needed columns
stats_df.drop([#'index',
 #'Season',
 #'Tm',
 #'G',
 #'W',
 #'L',
 #'Win_pct',
 #'MOV',
 #'SOS',
 #'SRS',
 'Pace',
 #'ORtg',
 #'DRtg',
 'eFG_pct',
 'TS_pct',
 'TOV_pct',
 'ORB_pct',
 'FT_FGA',
 'Opp_eFG_pct',
 'Opp_TS_pct',
 'Opp_TOV_pct',
 'Opp_ORB_pct',
 'Opp_FT_FGA'
              ], axis=1, inplace=True)
stats_df.head()

Position  Win_pct   MOV   SOS   SRS   ORtg   DRtg
0         0    0.472 -1.83 -0.01 -1.85  111.2  113.0
1         0    0.722  9.25 -0.29  8.97  117.6  108.3
2         0    0.375 -0.47 -0.07 -0.54  112.0  112.5
3         0    0.458 -1.74  0.15 -1.58  111.0  112.8
4         0    0.431 -3.68  0.23 -3.45  113.6  117.2

In [19]:
# Remove Position target from features data
y = stats_df.Position
X = stats_df.drop(columns=["Position"])

# Split training/test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)

In [20]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Scale the data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)

In [21]:
# Create a logistic regression model
classifier_SFM = LogisticRegression(max_iter=1000)
classifier_SFM

LogisticRegression(max_iter=1000)

In [22]:
# Fit (train) our model by using the training data
classifier_SFM.fit(X_train_scaled, y_train)

LogisticRegression(max_iter=1000)

In [23]:
# Validate the model by using the test data
print(f"Logistic Regression with SFM features only Training Data Score: {classifier_SFM.score(X_train_scaled, y_train)}")
print(f"Logistic Regression with SFM features only Testing Data Score: {classifier_SFM.score(X_test_scaled, y_test)}")

Logistic Regression with SFM features only Training Data Score: 0.9656652360515021
Logistic Regression with SFM features only Testing Data Score: 0.9743589743589743


SFS feature score = 0.974358

In [24]:
# Pull out Season, TM and Team_Name into DF
test_name_df = test_df[['Season', 'Tm']]
test_name_df.head()

Season   Tm
0  2021-22  MIA
1  2021-22  WAS
2  2021-22  UTA
3  2021-22  TOR
4  2021-22  SAS

In [25]:
test_df_pos= test_df['Position']
test_df.drop(['index',
 'Position',
 'Season',
 'Tm',
 'G',
 'W',
 'L',
 #'Win_pct',
 #'MOV',
 #'SOS',
 #'SRS',
 'Pace',
 #'ORtg',
 #'DRtg',
 'eFG_pct',
 'TS_pct',
 'TOV_pct',
 'ORB_pct',
 'FT_FGA',
 'Opp_eFG_pct',
 'Opp_TS_pct',
 'Opp_TOV_pct',
 'Opp_ORB_pct',
 'Opp_FT_FGA'
             ], axis=1, inplace=True)
test_df.head()

Win_pct    MOV   SOS    SRS   ORtg   DRtg
0    0.833  17.00  1.07  18.07  112.2   95.5
1    0.714   2.57  1.11   3.68  108.8  106.3
2    0.833  12.83 -3.08   9.76  111.9   99.0
3    0.625   4.63  1.25   5.88  107.4  102.7
4    0.286  -0.43 -1.50  -1.93  106.6  107.0

In [26]:
df_X_test_scaled = scaler.fit_transform(test_df)

In [27]:
df_test_df=classifier_SFM.predict(df_X_test_scaled)
df_test_df

array([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0], dtype=int64)

In [28]:
cy_test_df = pd.DataFrame(df_test_df, columns = ['Position'])
cy_test_df.head()

Position
0         1
1         0
2         0
3         0
4         0

In [29]:
joined_df = test_name_df.join(cy_test_df)
joined_df

Season   Tm  Position
0   2021-22  MIA         1
1   2021-22  WAS         0
2   2021-22  UTA         0
3   2021-22  TOR         0
4   2021-22  SAS         0
5   2021-22  SAC         0
6   2021-22  POR         0
7   2021-22  PHO         0
8   2021-22  PHI         0
9   2021-22  ORL         0
10  2021-22  OKC         0
11  2021-22  NYK         0
12  2021-22  NOP         0
13  2021-22  MIN         0
14  2021-22  MIL         0
15  2021-22  ATL         0
16  2021-22  MEM         0
17  2021-22  LAL         0
18  2021-22  LAC         0
19  2021-22  IND         0
20  2021-22  HOU         0
21  2021-22  GSW         0
22  2021-22  DET         0
23  2021-22  DEN         0
24  2021-22  DAL         0
25  2021-22  CLE         0
26  2021-22  CHO         0
27  2021-22  CHI         0
28  2021-22  BRK         0
29  2021-22  BOS         0

In [30]:
classifier_SFS.predict_proba(df_X_test_scaled)

NameError: name 'classifier_SFS' is not defined